Update cert-manager dependency to v1.4.0 (#16)
projectsyn/component-cert-manager: docs/modules/ROOT/pages/how-tos/upgrade.adoc

- The User-Agent request header sent by cert-manager has changed to reflect the ownership transfer to the CNCF
- The --renew-before-expiration-duration flag of the cert-manager controller-manager has been deprecated. Please set the Certificate.Spec.RenewBefore field instead. This flag will be removed in the next release.
Certificates issued by the Vault issuer have changed — the root CA instead of the issuing CA is now stored in ca.cr

https://github.com/jetstack/cert-manager/releases/tag/v1.3.0

- CertificateRequests are now immutable - the spec and metadata.annotations fields cannot be changed after creation. They were always designed to be immutable but this behavior is now enforced by the cert-manager webhook.
- This release updates the Venafi Cloud Issuer to use OutagePREDICT instead of DevOpsACCELERATE.

https://github.com/jetstack/cert-manager/releases/tag/v1.4.0

- The CA issuer now attempts to store the root CA instead of the issuing CA into the ca.crt field for issued certificates; this is a change of behavior. All of the information which was previously available is still available: the intermediate should appear as part of the chain in tls.crt.

> - RunAsNonRoot is now enabled by default in the securityContext values. If you're using custom containers with the chart that run as root, you will need to set this back to false.

Reviewer: It's not necessary to copy-paste the entire changelog of cert-manager itself. It's enough to link to the changelog upstream I'd say.

= How to Upgrade cert-manager

> == Upgrade cert-manager from 0.x to 1.x

Reviewer: Let's make each major version upgrade its own dedicated page
```suggestion
= Upgrade cert-manager from 0.x to 1.x
```

The upgrade is done with the following steps:

> . Prepare 3-way merge
> . Build Commodore catalog with updated `component-cert-manager`
> . Ensure ArgoCD sync
> . Verify certificates

Reviewer: ```suggestion
. Prepare CRD upgrade
. Compile Commodore catalog with `component-cert-manager` v2.x
. Verify cert-manager upgrade
```

> === Prerequisites

Reviewer: Naturally each child section goes "1 up"
```suggestion
== Prerequisites
```

. `commodore`
. `kubectl` or `oc`
. https://github.com/mikefarah/yq[`yq`] version 4 (alternatively, any editor works)

> === Prepare 3-way merge

Reviewer: The technical reason for a 3-way merge is not relevant, rather what it's about
```suggestion
== Prepare CRD upgrade
```

This step updates the `"kubectl.kubernetes.io/last-applied-configuration"` annotation which is required for a successful 3-way merge.

[source,bash]
----
cat <<'EOF' > cert-manager-prepare-merge-crds.sh
#!/bin/sh

set -euo pipefail

for crd in certificaterequests.cert-manager.io certificates.cert-manager.io challenges.acme.cert-manager.io clusterissuers.cert-manager.io issuers.cert-manager.io orders.acme.cert-manager.io; do
  # Get CRD definition in YAML
  kubectl get crd "${crd}" -o yaml > "${crd}.yaml"
  # Remove all metadata properties except `metadata.name`
  yq -i eval 'del(.status) | del(.metadata) | .metadata.name += "'${crd}'"' "${crd}.yaml"
  # Apply the CRD again (this shouldn't change anything, except adding the annotation "kubectl.kubernetes.io/last-applied-configuration")
  # You will also see some warnings in the output mentioning the annotation.
  # This is expected and actually required.
  kubectl apply -f "${crd}.yaml"
done
EOF

sh cert-manager-prepare-merge-crds.sh
----

> === Build Commodore catalog with updated `component-cert-manager`
> 
> [source,bash]
> ----
> # Set the cluster you'd like to upgrade
> CLUSTER_ID=XXX
> 
> # Build catalog and push changes
> commodore catalog compile $CLUSTER_ID --push -i
> ----

Reviewer: This is not a how-to-commodore, I'd say we keep it simple.
```suggestion
. Compile and push cluster catalog
----
```

> === Ensure ArgoCD sync
> 
> Extract the password for the webgui (replace pbcopy with something that works for your OS)
> 
> [source,bash]
> ----
> kubectl -n syn get secret/steward -o jsonpath='{.data.token}' | base64 --decode | pbcopy
> ----
> 
> Start a port-forwarding to access the ArgoCD web gui.
> 
> [source,bash]
> ----
> kubectl -n syn port-forward svc/argocd-server 8080:80
> ----
> 
> Visit http://localhost:8080/applications/cert-manager and click `🔄 SYNC`.

Reviewer: This step can be omitted. It's kind of a prerequisite to know how to operate ArgoCD.
Also, with a regular catalog compile + push, it's not necessary to explicitly click on Sync, Argo does it automatically when there's a new commit.
```suggestion
```

=== Verify cert-manager and certificates

> [source,bash]
> ----
> # Verify successful sync and health status in argo
> 
> # Check for errors in the logs
> kubectl -nsyn-cert-manager logs -l "app.kubernetes.io/name=cert-manager"

Reviewer: ```suggestion
. Verify successful sync and health status in ArgoCD
. Check for obvious deployment errors
+
[source,bash]
----
kubectl -n syn-cert-manager get pods
kubectl -n syn-cert-manager logs -l "app.kubernetes.io/name=cert-manager"
```

> # TODO can we / should we check certificates?

Author: Can and should we add some more checks?

Reviewer: It's an extra optional effort. One would need to have a DNS record and point to the cluster, then we can create an Ingress, then verify if the stack works, then remove again.

We could manually test the upgrade on a "lesser-important" cluster first where wildcard DNS is available. If that works, I'd consider the upgrade successful for all other clusters too.

Reviewer: Actually, maybe we should suggest it. It's up to the user whether they do it or not

Author: How do we suggest it? Keep it simple and suggest creating a new Ingress or some more guidance?

Reviewer: Check out the help page `kubectl create ingress --help`, there are a few rules described how to quickly generate one. The ingress doesn't need to resolve to a service + app, just a valid certificate when doing a `curl` request or so.
